### PR TITLE
texlive: generate "fixed-hashes.nix" from Nixpkgs attribute

### PR DESCRIPTION
**Pull request by xworld21**
###### Description of changes
Generate `fixed-hashes.nix` from a Nix expression. Necessary, because the old instructions do not work after transitioning to `tlDeps`: `texlive.scheme-full.pkgs` does not contain all packages any more. Moreover, the new expression allows generating the hashes of the *new* packages only, saving a lot of time during upgrades, and the process can be safely interrupted and restarted at any time without losing work.

###### Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by apfelkuchen6** [comment]
What is still missing here? I've rebased this on current master (with some slight changes to account for the new tlpkg packages) and it appears to work great: https://github.com/NixOS/nixpkgs/compare/master...apfelkuchen6:nixpkgs:texlive-nix-fixed-hashes

**Comment by xworld21** [comment]
> What is still missing here?

My time :( I'll catch up this weekend!

File: pkgs/tools/typesetting/tex/texlive/generate-fixed-hashes.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  # name + revision for the fixed output hashes
+  mkFixedName = { revision, extraRevision ? "", ... }@attrs: mkURLName attrs + ".r${toString revision}${extraRevision}";
+
+  uniqueByName = fods: catAttrs "fod" (genericClosure {
+    startSet = map (fod: { key = fod.name; inherit fod; }) fods;
+    operator = _: [ ];
+  });
+
+  # ugly hack to extract combine from collection-latexextra, since it is masked by texlive.combine
+  combine = lib.findFirst (p: (lib.head p.pkgs).pname == "combine") { pkgs = []; } (lib.head texlive.collection-latexextra.pkgs).tlDeps;
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/generate-fixed-hashes.nix:16]: Issue already reported in #237227.

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 2]
```diff
@@ -155,13 +154,24 @@ let
...
   '';
 
+  # map: name -> fixed-output hash
+  fixedHashes = lib.optionalAttrs useFixedHashes (import ./fixed-hashes.nix);
+
+  # NOTE: the fixed naming scheme must match generated-fixed-hashes.nix
+  # name for the URL
+  mkURLName = { pname, tlType, ... }: pname + lib.optionalString (tlType != "run" && tlType != "tlpkg") ".${tlType}";
+  # name + revision for the fixed output hashes
+  mkFixedName = { revision, extraRevision ? "", ... }@attrs: mkURLName attrs + ".r${toString revision}${extraRevision}";
```
Comment (apfelkuchen6) [inline pkgs/tools/typesetting/tex/texlive/default.nix:164]: this picks the same name for "run" and for "tlpkg" packages, which ultimately results in the tlpkg packages being replaced by dupicates of the run packages.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix]: Oh my, I forgot the `tlpkg` in all the *other* relevant places. Fixed now, I hope (they popped up in the new `fixed-hashes.nix` and I get exceptions if I omit one, so it should be good).

**Review by apfelkuchen6 — APPROVED** [review]
Looks great. This not changing the evaluation is also a really good sign.

**Pull request merged**